sycl/include/CL/sycl/exception.hpp

```diff
@@ -12,40 +12,46 @@
 
 #include <CL/sycl/detail/common.hpp>
 #include <CL/sycl/stl.hpp>
-#include <exception>
+
+#include <cstddef>
 
 namespace cl {
 namespace sycl {
 
+// Forward declaration
 class context;
 
 struct exception {
+
+  using ContextPtr = shared_ptr_class<context>;
+
   exception() = default;
 
-  const char *what() const noexcept { return msg.c_str(); }
+  exception(const char *Msg, const cl_int ClErr = CL_SUCCESS,
```

Comment:
Could you change semantic of c'tor so it takes usual sycl::context type and store sycl::detail::context_impl inside?